cpp/src/io/parquet/writer_impl_helpers.cpp

```diff
@@ -0,0 +1,138 @@
+      return Compression::LZ4_RAW;
+    case compression_type::NONE: return Compression::UNCOMPRESSED;
+    default: CUDF_FAIL("Unsupported compression type");
+  }
+}
+
+nvcomp::compression_type to_nvcomp_compression_type(Compression codec)
+{
+  if (codec == Compression::SNAPPY) return nvcomp::compression_type::SNAPPY;
+  if (codec == Compression::ZSTD) return nvcomp::compression_type::ZSTD;
+  // Parquet refers to LZ4 as "LZ4_RAW"; Parquet's "LZ4" is not standard LZ4
+  if (codec == Compression::LZ4_RAW) return nvcomp::compression_type::LZ4;
+  CUDF_FAIL("Unsupported compression type");
+}
+
+uint32_t page_alignment(Compression codec)
+{
+  if (codec == Compression::UNCOMPRESSED or
+      nvcomp::is_compression_disabled(to_nvcomp_compression_type(codec))) {
+    return 1u;
+  }
+
+  return 1u << nvcomp::compress_input_alignment_bits(to_nvcomp_compression_type(codec));
+}
+
+size_t max_compression_output_size(Compression codec, uint32_t compression_blocksize)
+{
+  if (codec == Compression::UNCOMPRESSED) return 0;
+
+  return compress_max_output_chunk_size(to_nvcomp_compression_type(codec), compression_blocksize);
+}
+
+void fill_table_meta(std::unique_ptr<table_input_metadata> const& table_meta)
+{
+  // Fill unnamed columns' names in table_meta
+  std::function<void(column_in_metadata&, std::string)> add_default_name =
+    [&](column_in_metadata& col_meta, std::string default_name) {
+      if (col_meta.get_name().empty()) col_meta.set_name(default_name);
+      for (size_type i = 0; i < col_meta.num_children(); ++i) {
+        add_default_name(col_meta.child(i), col_meta.get_name() + "_" + std::to_string(i));
+      }
+    };
+  for (size_t i = 0; i < table_meta->column_metadata.size(); ++i) {
+    add_default_name(table_meta->column_metadata[i], "_col" + std::to_string(i));
+  }
+}
+
+[[nodiscard]] size_t column_size(column_view const& column, rmm::cuda_stream_view stream)
+{
+  if (column.is_empty()) { return 0; }
+
+  if (is_fixed_width(column.type())) {
+    return size_of(column.type()) * column.size();
+  } else if (column.type().id() == type_id::STRING) {
+    auto const scol = strings_column_view(column);
+    return cudf::strings::detail::get_offset_value(
+             scol.offsets(), column.size() + column.offset(), stream) -
+           cudf::strings::detail::get_offset_value(scol.offsets(), column.offset(), stream);
+  } else if (column.type().id() == type_id::STRUCT) {
+    auto const scol = structs_column_view(column);
+    size_t ret      = 0;
+    for (int i = 0; i < scol.num_children(); i++) {
+      ret += column_size(scol.get_sliced_child(i, stream), stream);
+    }
+    return ret;
+  } else if (column.type().id() == type_id::LIST) {
+    auto const lcol = lists_column_view(column);
+    return column_size(lcol.get_sliced_child(stream), stream);
+  }
+
+  CUDF_FAIL("Unexpected compound type");
+}
+
+[[nodiscard]] bool is_col_nullable(cudf::detail::LinkedColPtr const& column,
```

Comment:
The function should technically work for input columns `column_view` instead of `cudf::detail::LinkedColPtr` as well since `column->null_count()` is available in both classes. We could make it a template as well but I don't think it's being used elsewhere yet.

Reply:
In short, I don't mind it being renamed to `is_output_column_nullable` so either or is good!

Reply:
honestly, it's only used for schema construction...it would probably be better to just leave it in `writer_impl.cu` or turn it into a lambda in `construct_schema_tree()`.

Reply:
the reason I'm suggesting this is because AFAICT this function determines, from the input column and writer options, if the written column should be nullable. It's logic specific to the PQ writer (again, IIUC).

Reply:
You are right but looking at this function as-is, it takes in an arbitrary column view and returns its nullability. The returned nullability in our case is used to set the corresponding the output column, yes, but the function can be used outside this context as well if ever needed.

Reply:
All in all, I don't have a strong preference one way or the other (happy to do it if folks like) but was just putting my thoughts here.

Reply:
The only thing that keeps it from being general is the bit about `write_mode == single_write_mode::NO`. Given this is used only to set the nullability in the output schema, it's probably best to use @vuule's suggestion to avoid future confusion.

Reply:
Sounds good. Makes sense!